File: extract.js
const { Lens } = require('chrome-lens-ocr');
const fs = require('fs');
const path = require('path');
const { fromPath } = require('pdf2pic'); // لتحويل الـ PDF لصور

async function processPdfs() {
    const lens = new Lens('ar'); // تحديد اللغة العربية
    const inputDir = './pdfs';
    const outputDir = './Extracted_Texts';

    if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir);

    const files = fs.readdirSync(inputDir).filter(f => f.endsWith('.pdf'));

    for (const file of files) {
        console.log(`Processing: ${file}`);
        const options = {
            density: 300,
            saveFilename: "temp_page",
            savePath: "./temp",
            format: "png",
            width: 2000,
            height: 2000
        };
        
        if (!fs.existsSync("./temp")) fs.mkdirSync("./temp");
        
        const storeAsImage = fromPath(path.join(inputDir, file), options);
        // تحويل أول صفحة (أو عدل الحلقة لتشمل كل الصفحات)
        const pageRes = await storeAsImage(1); 

        try {
            const result = await lens.scanByFile(pageRes.path);
            const text = result.segments.map(s => s.text).join(' ');
            
            fs.writeFileSync(path.join(outputDir, file + '.txt'), text);
            console.log(`✅ تم استخراج النص من ${file}`);
        } catch (e) {
            console.error(`❌ خطأ في ${file}:`, e.message);
        }
    }
}

processPdfs();

File: extract.rb
require 'google-lens-ocr'
require 'fileutils'

# إعداد المجلدات
input_dir = './pdfs'
output_dir = './Extracted_Texts'
temp_dir = './temp'

FileUtils.mkdir_p(output_dir)
FileUtils.mkdir_p(temp_dir)

# البحث عن ملفات PDF
pdf_files = Dir.glob("#{input_dir}/*.pdf")

pdf_files.each do |pdf_path|
  puts "Processing: #{File.basename(pdf_path)}..."
  
  # اسم ملف الصورة المؤقتة
  temp_image = "#{temp_dir}/page.png"
  
  # تحويل الصفحة الأولى من PDF إلى صورة باستخدام pdftoppm (موجود في poppler-utils)
  system("pdftoppm -f 1 -l 1 -png -singlefile '#{pdf_path}' '#{temp_dir}/page'")
  
  begin
    # استخدام Google Lens لاستخراج النص
    # الأداة تتعرف على اللغة العربية تلقائياً
    lens = GoogleLensOCR.new
    result = lens.extract_text(temp_image)
    
    # حفظ النص المستخرج
    output_file = File.join(output_dir, File.basename(pdf_path, ".pdf") + ".txt")
    File.write(output_file, result)
    
    puts "✅ Successfully extracted: #{pdf_path}"
  rescue => e
    puts "❌ Error processing #{pdf_path}: #{e.message}"
  ensure
    # تنظيف الصورة المؤقتة
    FileUtils.rm(temp_image) if File.exist?(temp_image)
  end
end
